core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/Collector.scala

```diff
@@ -170,5 +177,38 @@ trait Collector {
         unsecureEndpoint(routes)
         ()
     }
+
+    Runtime
+      .getRuntime
+      .addShutdownHook(new Thread(() => {
+        log.warn("Received shutdown signal")
+        if (collectorConf.preTerminationUnhealthy) {
+          log.warn("setting health endpoint to unhealthy")
+          health.toUnhealthy()
+        }
+        log.warn(s"Sleeping for ${collectorConf.preTerminationPeriod}")
+        Thread.sleep(collectorConf.preTerminationPeriod.toMillis)
+        log.warn("Initiating http server termination")
+        try {
+          // The actor system is already configured to shutdown within `terminationDeadline` so we await for double that.
+          Await.result(system.terminate(), collectorConf.terminationDeadline * 2)
+          log.warn("Server terminated")
+        } catch {
+          case NonFatal(t) =>
+            log.error("Caught exception awaiting server termination", t)
+        }
+        log.warn("Initiating sinks shutdown")
+        val shutdowns = List(shutdownSink(sinks.good), shutdownSink(sinks.bad))
+        Await.result(Future.sequence(shutdowns), Duration.Inf)
+        ()
+      }))
   }
+
+  private def shutdownSink(sink: Sink)(implicit ec: ExecutionContext): Future[Unit] =
+    Future {
+      sink.shutdown()
```

Comment:
Very last thing - when we look at the logs it'd be great to see that sink `A` has flushed successfully (although doesn't seem we can get meaningful names for them)

Reply:
I did what I could.  It now says "good sink" or "bad sink" before trying to terminate and after successfully terminating.